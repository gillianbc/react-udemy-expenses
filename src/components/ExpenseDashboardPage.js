// Stateless functional component
import React from "react";
import MyExpenses from "./ExpenseList";

const ExpenseDashboardPage = () => (
    <div>
        <MyExpenses />
    </div> )

export default ExpenseDashboardPage