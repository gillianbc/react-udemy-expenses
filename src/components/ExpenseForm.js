import React, {Component} from 'react';

class ExpenseForm extends Component {
    onChangeDescription = (e) => {
        const description  = e.target.value
        // Andrew used a function - read this for why https://www.freecodecamp.org/news/functional-setstate-is-the-future-of-react-374f30401b6b/
        this.setState( () => ({ description }))
        // Passing an object like so will work, but don't do it
        // this.setState({ description })
    }
    onChangeNote = (e) => {
        const note  = e.target.value
        this.setState(() => ({ note }))
    }
    onChangeAmount = (e) => {
        //   /^\d*(\.\d{0,2})?$
        const amount = e.target.value
        if (amount.match(/^\d*(\.\d{0,2})?$/)){
            this.setState(() => ({ amount }))
        }
    }
    state = {
        description: '',
        note: '',
        amount: '0.00'
    }
    render() {
        return (
            <div>
                <form>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value = {this.state.description}
                        onChange = {this.onChangeDescription}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value = {this.state.amount}
                        onChange = {this.onChangeAmount}
                    />
                    <textarea
                        placeholder="Enter an optional note for the expense"
                        value = {this.state.note}
                        onChange = {this.onChangeNote}
                    />
                    <button >Add Expense</button>

                </form>
            </div>
        );
    }
}

export default ExpenseForm;