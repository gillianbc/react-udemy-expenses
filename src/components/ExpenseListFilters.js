import React from 'react';
import {connect} from "react-redux";
import {setTextFilter, sortByAmount, sortByDate} from "../actions/filters";
class ExpenseListFilters extends React.Component {
    render(){
        return (<div>
            <input type="text" value={this.props.filters.text} onChange={ (e) => {
                this.props.dispatch(setTextFilter(e.target.value))
            }}/>
            <div>
                <select
                    value={this.props.filters.sortBy}
                    onChange={ (e) => {
                        e.target.value === "date" ? this.props.dispatch(sortByDate()) : this.props.dispatch(sortByAmount())
                    }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
            </div>

        </div>)
    }
}

// Grab data from the connected state and put it in the props
const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const ConnectedExpenseListFilters = connect(mapStateToProps)(ExpenseListFilters)

export default ConnectedExpenseListFilters