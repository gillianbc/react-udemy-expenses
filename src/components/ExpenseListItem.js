import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ id, amount, description, createdAt }) => {
  console.log("EXPENSE LIST ITEM", description);
  return (
    <div>
      <p>
        <Link to={"edit/" + id}>{description}</Link> Â£{amount} Created Date:{" "}
        {createdAt}.
      </p>
    </div>
  );
};

export default ExpenseListItem;
