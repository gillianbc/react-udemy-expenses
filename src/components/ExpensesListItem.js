import React from 'react'

export const ExpensesListItem = ({description, amount, createdAt}) => (
    <div>
        <p>Item: {description} Â£{amount} Created Date: {createdAt}</p>
    </div>
)