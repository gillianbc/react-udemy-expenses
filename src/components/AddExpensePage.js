// Stateless functional component
import React from "react";
import ExpenseForm from "./ExpenseForm";
import {connect} from "react-redux";
import {addExpense} from "../actions/expenses";

// We pass a callback function down as a prop to the ExpenseForm
// Let me explain the JSX syntax.  We need curlies for JS within JSX
// onSubmit={some javascript}
// (expense) => {}  is simply a single line arrow function
const AddExpensePage = (props) => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm
            onSubmit={(expense) => {
                console.log('Add Expense Page', props)
                console.log('Expense page received an expense from the expense form', expense)
                props.dispatch(addExpense(expense));
                /* This is how we redirect */
                props.history.push('/')
            }}
        />
    </div> )
const ConnectedAddExpensePage = connect()(AddExpensePage)
export default ConnectedAddExpensePage