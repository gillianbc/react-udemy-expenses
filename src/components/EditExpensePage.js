// Stateless functional component
import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { addExpense, removeExpense } from "../actions/expenses";
import { editExpense } from "../actions/expenses";

const EditExpensePage = (props) => (
  <div>
    <h1>Edit Expense</h1>
    <ExpenseForm
      expense={props.expense}
      onSubmit={(expense) => {
        expense.id
          ? props.dispatch(editExpense(expense.id, expense))
          : props.dispatch(addExpense(expense));
        /* This is how we redirect.  history is always part of the props */
        props.history.push("/");
      }}
    />
    <button
      onClick={() => {
        props.dispatch(removeExpense({ id: props.expense.id }));
        props.history.push("/");
      }}
    >
      Remove
    </button>
  </div>
);

// We need to pass the id in from our class's props and get an expense object out of the store (aka the global state)
// If we then examine the props coming in, we will see that it has then been populated with the expense from the state
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    ),
  };
};

// connect to the store
export default connect(mapStateToProps)(EditExpensePage);
