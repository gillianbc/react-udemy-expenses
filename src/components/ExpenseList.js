import React from 'react'
import {connect} from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import {getVisibleExpenses} from "../selectors/expenses";

const ExpenseList = (props) => (
    <div>
        <h1>This is my Expense List</h1>
        <p>My name is {props.gillian}</p>
        <p>The text filter from the state is: {props.filters.text}</p>
        {props.expenses.map((expense) => {
            // When we map the array of expenses, each row is an object
            console.log('Processing in map', expense)
            // Send across an object consisting of each field of the expense object
             return <ExpenseListItem key={expense.id} {...expense}/>
            // The above sends across key and each items of expense to props (or whatever we decide to call the args object)
            // { id: "jskldjkla", amount: 1888 } etc
            // We can then destructure
            // Note that key is not one of the ordinary props - it's a special keyword for React to use for array row handling
            // Alternately, I could have done:
            // return <ExpenseListItem key={expense.id} expense={expense}/>
            // The above sends across
            // { expense: {id: "jskldjkla", amount: 1888 } } etc
            // Alternately, I could have done this which is the preferred form when only a few properties as it's
            // explicit and you only pass what you need:
            //return <ExpenseListItem key={expense.id}  amount={expense.amount} description={expense.description}/>
            // Remember that the {} here are not object brackets, they are so we can write JS within the JSX
        })}
    </div>
)

// This is just a function that returns an object of the things we may want to use that are in our application's state
// Note: declare it before we need to use it
const mapStateToProps = (state) => {
    console.log('Expenses === ', state.expenses)
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters),
        gillian: 'Mrs Bladen-Clark',
        filters: state.filters
    }
}

// This does look like strange syntax, but it's simple really.
// connect() returns a function and we then call that function with our ExpenseList component
// just like we did with WrappedComponent in hoc.js
const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)

// Note that we now export the connected expense list
export default ConnectedExpenseList;