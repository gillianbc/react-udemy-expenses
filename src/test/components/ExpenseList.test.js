import React from 'react'
import { ExpenseList } from '../../components/ExpenseList'
import { shallow } from 'enzyme'
import testExpenseData from '../resources/expenses'
/*
* Tests the snapshot of the ExpenseList component
* */

it('The snapshot should match when there is an array of expenses', () => {
    // Remember, these curlies are not object brackets, they're JSX brackets
    const wrapper = shallow(<ExpenseList expenses={ testExpenseData }/>);
    expect(wrapper).toMatchSnapshot();
})

it('The snapshot should match when there is an empty array of expenses', () => {
    // Remember, these curlies are not object brackets, they're JSX brackets
    const wrapper = shallow(<ExpenseList expenses={ [] }/>);
    expect(wrapper).toMatchSnapshot();
})

