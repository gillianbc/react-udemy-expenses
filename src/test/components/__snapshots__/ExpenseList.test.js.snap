// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The snapshot should match when there is an array of expenses 1`] = `
<div>
  <h1>
    This is my Expense List
  </h1>
  <ExpenseListItem
    amount={196}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={3456000000}
    description="Credit Card"
    id="3"
    key="3"
    note="I never pay interest as I clear the whole balance every month"
  />
</div>
`;

exports[`The snapshot should match when there is an empty array of expenses 1`] = `
<div>
  <h1>
    This is my Expense List
  </h1>
</div>
`;
